### PR TITLE
generate lax.slice instead of lax.gather for x[<int>] or x[:<int>]

### PR DESCRIPTION
**Pull request by mattjj**
Before, on CPU:
image
image

After, on CPU:
image
image

Following up on #11866.

I don't know if these small compile time differences matter, but hey, roofshots! And the jaxpr pretty-printing win is enough for me.

### CODE REVIEW

File: jax/_src/numpy/lax_numpy.py [hunk 1]
```diff
@@ -3594,6 +3595,23 @@ def _rewriting_take(arr, idx, indices_are_sorted=False, unique_indices=False,
...
+    if 0 <= idx < len(arr):
+      return lax.index_in_dim(arr, idx, keepdims=False)
+  if (isinstance(idx, slice) and
+      (type(idx.start) is int or idx.start is None) and
+      (type(idx.stop)  is int or idx.stop is  None) and
+      (type(idx.step)  is int or idx.step is  None)):
+    n = len(arr)
+    start = idx.start if idx.start is not None else 0
+    stop  = idx.stop  if idx.stop  is not None else n
+    step = idx.step if idx.step is not None else 1
```
Comment (froystig) [inline jax/_src/numpy/lax_numpy.py:3610]:
```suggestion
    start = idx.start if idx.start is not None else 0
    stop  = idx.stop  if idx.stop  is not None else n
    step  = idx.step  if idx.step  is not None else 1
```

**Review by froystig — APPROVED** [review]
Nice work!

File: jax/_src/numpy/lax_numpy.py [hunk 2]
```diff
@@ -82,6 +82,7 @@
 from jax.experimental.array import Array
 
 newaxis = None
+NoneType = type(None)
```
Comment (sharadmv) [inline jax/_src/numpy/lax_numpy.py:85]: What's this for?

**Review by sharadmv — APPROVED** [review]

File: jax/_src/numpy/lax_numpy.py (hunk 2 above)
Comment (mattjj) [reply jax/_src/numpy/lax_numpy.py:85]: Oops, stale! (I was doing `isinstance(idx.step, NoneType)` etc., but changed when mypy complained.)

**Pull request merged**